analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs

```diff
@@ -0,0 +1,41 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers
+{
+    public sealed partial class KnownReference
+    {
+        private readonly Func<IEnumerable<AssemblyIdentity>, bool> predicate;
+
+        internal KnownReference(Func<AssemblyIdentity, bool> predicate) : this(Any(predicate))
+        {
+        }
+
+        internal KnownReference(Func<IEnumerable<AssemblyIdentity>, bool> predicate)
+        {
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+        }
+
+        public static KnownReference XUnit_Assert { get; } = new(NameIs("xunit.assert").Or(NameIs("xunit").And(VersionLowerThen("2.0"))));
```

Comment:
I created a small DSL to define the references. This allows for unifying the checks which should improve maintainability (e.g. the Name, Version, and PublicKey checks might be buggy).

XUnit extracted the "Assert" part in version 2 into a separate assembly so the check is
* "xunit.assert" or
* "xunit" and Version before 2.0

Something like this could also be used for e.g. "SqlClient" for Microsoft SQL server which is distributed in several different assemblies [System.Data.SqlClient](https://www.nuget.org/packages/System.Data.SqlClient) and [Microsoft.Data.SqlClient](https://www.nuget.org/packages/Microsoft.Data.SqlClient)

Reply:
It's nice, but somewhat hard to follow with all the possibilities. We'd typically need A || (B && C) || (D && E && F) schema.
As a first step, let's remove `Or`, keep `And`, and make it `params` that will behave as "any". That should simplify it.

```suggestion
        public static KnownReference XUnit_Assert { get; } = new(
            NameIs("xunit.assert"), 
            NameIs("xunit").And(VersionLowerThen("2.0"))));
```